include/mfmg/adapters_dealii.hpp

```diff
@@ -0,0 +1,199 @@
+  {
+    using matrix_type = typename local_operator_type::matrix_type;
+    using sparsity_pattern_type =
+        typename local_operator_type::sparsity_pattern_type;
+
+    auto system_sparsity_pattern = std::make_shared<sparsity_pattern_type>();
+    auto system_matrix = std::make_shared<matrix_type>();
+
+    // Call user function to fill in the matrix and build the mass matrix
+    evaluate(mesh._dof_handler, mesh._constraints, *system_sparsity_pattern,
+             *system_matrix);
+
+    return std::make_shared<local_operator_type>(system_matrix,
+                                                 system_sparsity_pattern);
+  }
+
+  // For deal.II, because of the way it deals with hanging nodes and
+  // Dirichlet b.c., we need to zero out the initial guess values
+  // corresponding to those. Otherwise, it may cause issues with spurious
+  // modes and some scaling difficulties. If we use `apply` for deal.II, we
+  // don't need this as we can apply the constraints immediately after
+  // applying the matrix and before any norms and dot products.
+  //
+  // In addition, this has to work only for ARPACK with dealii::Vector<double>
+  void set_initial_guess(const mesh_type &mesh,
+                         typename local_operator_type::vector_type &x) const
+  {
+    unsigned int const n = x.size();
+
+    std::default_random_engine generator;
+    std::uniform_real_distribution<value_type> distribution(0., 1.);
+    for (unsigned int i = 0; i < n; ++i)
+      x[i] = (mesh._constraints.is_constrained(i) == false
+                  ? distribution(generator)
+                  : 0.);
+  }
+};
+
+template <int dim, class VectorType>
+class Adapter<DealIIMeshEvaluator<dim, VectorType>>
+{
+protected:
+  using mesh_evaluator_type = DealIIMeshEvaluator<dim, VectorType>;
+  using mesh_type = typename mesh_evaluator_type::mesh_type;
+  using operator_type = typename mesh_evaluator_type::operator_type;
+  using global_operator_type =
+      typename mesh_evaluator_type::global_operator_type;
+  using smoother_type = DealIISmootherOperator<VectorType>;
+  using direct_solver_type = DealIIDirectOperator<VectorType>;
+  using vector_type = typename operator_type::vector_type;
+
+public:
+  static std::shared_ptr<operator_type>
+  build_restrictor(MPI_Comm comm, const mesh_evaluator_type &evaluator,
+                   const mesh_type &mesh,
+                   std::shared_ptr<boost::property_tree::ptree> params)
+  {
+    AMGe_host<mesh_type::dimension(), mesh_evaluator_type, vector_type> amge(
+        comm, mesh._dof_handler);
+
+    std::array<unsigned int, dim> agglomerate_dim;
+    agglomerate_dim[0] = params->get<unsigned int>("agglomeration: nx");
+    agglomerate_dim[1] = params->get<unsigned int>("agglomeration: ny");
+    if (dim == 3)
+      agglomerate_dim[2] = params->get<unsigned int>("agglomeration: nz");
+    int n_eigenvectors =
+        params->get<int>("eigensolver: number of eigenvectors", 1);
```

Comment:
I guess I just prefer flat lists, as it requires a single step of getting a stored value out, rather than first getting sublist and then getting the value.

Reply:
You can also do `param->get("eingensolver.number of eigenvectors", 1);" if you prefer. You don't need the get the subtree first.

Reply:
Oh, did not know that. So that also means that one cannot use `.` inside the parameter names?

Reply:
Good question, I have never tried. It might work but then the part on the right of `.` would be the real name of the parameter and on the left you would have the name of the "sub-tree"